tests/unit/supplements/factory_supplements_test.py

```diff
@@ -41,35 +41,76 @@ def test_register_conflicting_key():
 
-    assert get_user_defined_mapping()['a']() == 1
+    assert get_user_defined_mapping()['a']['opid']() == 1
+
+
+class TestEndpointFixtures:
+    """
+    Note: Since defaultdict creates a key-value pair once a non-existent key
+    is called, if we change the default value, keys that have been encountered
+    will not have the new value. This shouldn't be a problem since we don't
+    do not request keys until all the factories are registered.
+    """
+
+    def test_register_endpoints_default(self):
+        register_function('a', return_value=1)
+
+        assert get_user_defined_mapping()['a']['endpoint']() == 1
+        assert get_user_defined_mapping()['a']['other_endpoint']() == 1
+
+    def test_register_endpoints_no_default(self):
+        register_function('a', endpoints=['endpoint'], return_value=1)
+
+        assert get_user_defined_mapping()['a']['endpoint']() == 1
+        assert get_user_defined_mapping()['a']['other_endpoint']() is None
+
+    def test_register_endpoints_both(self):
+        register_function('a', return_value=1)
+        register_function('a', endpoints=['other_endpoint'], return_value=2)
+
+        assert get_user_defined_mapping()['a']['endpoint']() == 1
+        assert get_user_defined_mapping()['a']['other_endpoint']() == 2
 
 
 class TestInjectVariables:
 
     def setup(self):
         fuzz_lightyear.register_factory('nested_dependency')(self.nested_dependency)
         fuzz_lightyear.register_factory('caller')(self.caller)
+        fuzz_lightyear.register_factory('caller', 'new_opid')(self.special_caller)
+        fuzz_lightyear.register_factory('caller', 'only_opid')(self.endpt_specific_caller)
         fuzz_lightyear.register_factory('dependency')(self.dependency)
+        fuzz_lightyear.register_factory('endpt_dependency', 'only_opid')(self.dependency)
 
     def test_uses_default(self):
-        assert get_user_defined_mapping()['caller']() == 2
+        assert get_user_defined_mapping()['caller']['opid']() == 2
+
+    def test_nested_endpoint_dependency(self):
+        assert get_user_defined_mapping()['caller']['new_opid']() == 3
 
     def test_throws_error_when_no_default(self):
         def foobar(no_default):
             pass
         fuzz_lightyear.register_factory('a')(foobar)
 
         with pytest.raises(TypeError):
-            get_user_defined_mapping()['a']()
+            get_user_defined_mapping()['a']['opid']()
 
+    # We should never use the endpoint-specific fixture.
     def test_nested_dependency(self):
-        assert get_user_defined_mapping()['nested_dependency']() == 4
+        assert get_user_defined_mapping()['nested_dependency']['opid']() == 4
+        assert get_user_defined_mapping()['nested_dependency']['new_opid']() == 4
+
+    # If the dependency is endpoint-specific then we should fail.
+    def test_endpoint_dependency_fails(self):
+        with pytest.raises(TypeError):
+            get_user_defined_mapping()['caller']['only_opid']()
```

Comment:
Huh. Why does this fail?

From the looks of things, it looks like this is the call stack:

```
operation_id == "only_opid". (caller, operation_id,) mapped to `endpt_specific_caller`
    calls `endpt_specific_caller`, which first loads `endpt_depedency`
        `endpt_dependency` mapped to `dependency`, for "only_opid". Since it is the same operation_id, it should work, and return 1.
    `endpt_specific_caller` returns 1 + 2
```

Reply:
This is a limitation of `inject_user_defined_variables`, since we would need to pass in the endpoint to the function from `register_factory` (see [here](https://github.com/Yelp/fuzz-lightyear/blob/8c262de83ae6609635f5edd2f6c2b809909f6f4c/fuzz_lightyear/supplements/factory.py#L94)) for it to be able to look at the mapping (used [here](https://github.com/Yelp/fuzz-lightyear/blob/8c262de83ae6609635f5edd2f6c2b809909f6f4c/fuzz_lightyear/datastore.py#L192)) to grab the right endpoint specified. ~~I think this would require a good amount more work (and might not be backwards compatible), since `inject_user_defined_variables` would need to be wrapped in another function for it to take arguments, and we'd need to change existing decorator references to `@inject_user_defined_variables()`. If you think there's a simpler way of doing this then I'd be happy to try it out.~~ I found [this](https://stackoverflow.com/a/3931903) as a way to work around this issue, so I believe that this should be doable.

Reply:
After some more finangling, I realized that fully implementing nested fixture dependency in combination with multiple specified endpoints is hard to do. Let's go through a couple of examples to show why.
Here's an easy example (which is allowed based on the current commit):
```
@register_factory('first_name', endpoint_ids='a')
def first_name():
    return 'john'
@register_factory('name', endpoint_ids='a')
def name(first_name):
    return first_name + ' doe'
```
This makes sense and is allowed. When we call name(), we [pass in](https://github.com/Yelp/fuzz-lightyear/blob/8fde87231201aa0ef345c6b384c48dfb32d5152b/fuzz_lightyear/supplements/factory.py#L92) the `operation_ids` specified in the decorator into `inject_user_defined_variables`. Inside `inject_user_defined_variables`, we take the `operation_id` specified and use it to [pass in](https://github.com/Yelp/fuzz-lightyear/blob/8fde87231201aa0ef345c6b384c48dfb32d5152b/fuzz_lightyear/datastore.py#L209) the right value based on the endpoint to value mapping. 

Now, let's take it a step further. What happens if we register `name` with two endpoints?
```
@register_factory('first_name', endpoint_ids='a')
def first_name():
    return 'john'
@register_factory('first_name', endpoint_ids='b')
def first_name():
    return 'bob'
@register_factory('name', endpoint_ids='a, b')
def name(first_name):
    return first_name + ' doe'
```
Logically, this would mean fuzzing endpoint `a` would use `john doe` as the name, and fuzzing `b` would return `bob doe`. However, when we try to get the value for `name` from the user defined mapping, we don't pass the endpoint value into the [wrapped function](https://github.com/Yelp/fuzz-lightyear/blob/8fde87231201aa0ef345c6b384c48dfb32d5152b/fuzz_lightyear/supplements/factory.py#L86). Now that I'm writing this, I think we might be able to handle this by always expecting and passing in a `operation_id` argument to the wrapper for the function, and then using that value to grab the right [nested value](https://github.com/Yelp/fuzz-lightyear/blob/8fde87231201aa0ef345c6b384c48dfb32d5152b/fuzz_lightyear/supplements/factory.py#L92) instead. I'm going to try this out - do let me know if you think this is a good idea.

Reply:
You might be able to just do:

```diff
88,94c88,89
<         # We only allow single operation ids to be passed through nested
<         # factories because we can't resolve many to one factories without
<         # rewriting how we handle endpoint specific fixtures.
<         if len(_operation_ids) == 1:
<             wrapped = inject_user_defined_variables(func, operation_id=_operation_ids[0])
<         else:
<             wrapped = inject_user_defined_variables(func, )
---
>         # Used when _operation_ids is not provided
>         wrapped = inject_user_defined_variables(func)
102c97
<                         mapping[key][operation_id] = wrapped
---
>                         mapping[key][operation_id] = inject_user_defined_variables(func, operation_id=operation_id)
113c108
<                             func_dict[operation_id] = wrapped
---
>                             func_dict[operation_id] = inject_user_defined_variables(func, operation_id=operation_id)
```

Reply:
You're right: this does solve the case that I described. However, this doesn't fix the test case `test_nested_endpoint_dependency_uses_default `. In our case, we have:
- A `nested_dependency` that is registered for generic endpoints, returning `caller * 2` 
- A `caller` that is registered for generic endpoints and `new_opid`, returning `dependency + 1` and `dependency + 2` respectively
- A `dependency` registered for generic endpoints that returns `1`.
Now, when we call  `nested_dependency` with `new_opid`, we get `4` instead of `6`. This is because `nested_dependency` is generic, so when we call `register_factory` it doesn't call `inject_user_defined_variables` with the list of `operation_ids`, since there is no list. So in conclusion, this fix works for specifying multiple endpoints and dependency injection, but users will not be able to use specific fixtures nested within generic fixtures. Personally I'm fine with leaving it like this - we can just take note of this situation and recommend to _always specify the endpoints that will be using a fixture_.

Reply:
That makes sense.

I'm fine with applying this patch, changing the test case to expect `6` with `xfail`, then creating an issue to describe this scenario in detail and referencing it in the `xfail` reason. Then we can merge this.